### PR TITLE
match gsn solidity version

### PR DESCRIPTION
**Pull request by dimalinux**
* Compile the gsnforwarder contract with solidity 0.8.7 (using docker) with optimization
* Pulled opengsn's source files from etherscan to generated byte-identical contract code (excluding metadata)
* Updates dependencies and go version to 1.20

### CODE REVIEW

File: .golangci.yml [hunk 1]
```diff
@@ -11,10 +11,22 @@ linters-settings:
...
     # report about shadowed variables
     check-shadowing: true
+    enable-all: true
+    disable:
+      - fieldalignment
+
+  revive:
+    rules:
+      - name: package-comments
+        disabled: true
```
Comment (dimalinux) [inline .golangci.yml:21]: I took the `golangci.yml` from `atomic-swap` and made 2 changes:
* Disabled the package-comments requirement (code at this comment)
* Removed  the tagliatelle checks for camel cased JSON tags

File: .golangci.yml [hunk 2]
```diff
@@ -11,10 +11,22 @@ linters-settings:
...
+
+  goimports:
+    local-prefixes: github.com/athanorlabs/atomic-swap
```
Comment (dimalinux) [inline .golangci.yml:24]: Note: update to atomic-relay

File: go.mod [hunk 3]
```diff
@@ -1,144 +1,153 @@
 module github.com/athanorlabs/go-relayer
 
-go 1.19
+go 1.20
 
 require (
 	github.com/athanorlabs/go-p2p-net v0.0.0-20230126054330-40336771ef86
```
Comment (dimalinux) [inline go.mod:6]: This repo is bit-rotting faster than we can maintain it. We'd have to make major updates to code we don't use anywhere to update this dependency.

File: impls/gsnforwarder/scripts/generate-bindings.sh [hunk 4]
```diff
@@ -1,36 +1,36 @@
...
-	--type IForwarder \
-	--out i_forwarder.go
+	--out forwarder.go
 
-sed '31,41d' tmp_forwarder.go > forwarder.go
-rm tmp_forwarder.go
\ No newline at end of file
+# IForwarderForwardRequest gets generated in both forwarder.go and i_forwarder.go
+# with an identical definition.
+sed -i 's/^type IForwarderForwardRequest /type _duplicate_IForwarderForwardRequest /' forwarder.go
```
Comment (dimalinux) [inline impls/gsnforwarder/scripts/generate-bindings.sh:36]: I assume this is solving the same problem that `sed` was solving in the previous version, but there wasn't any comment describing what it was doing, so I ended up reinventing the solution.

File: impls/gsnforwarder/scripts/generate-bindings.sh [hunk 5]
```diff
@@ -1,36 +1,36 @@
...
-
-"${SOLC_BIN}" --abi contracts/IForwarder.sol -o contracts/abi/ --overwrite
-"${SOLC_BIN}" --bin contracts/IForwarder.sol -o contracts/bin/ --overwrite
-
-"${ABIGEN}" \
-	--abi contracts/abi/IForwarder.abi \
-	--bin contracts/bin/IForwarder.bin \
-	--pkg ${PKG_NAME} \
-	--type IForwarder \
-	--out i_forwarder.go
```
Comment (dimalinux) [inline impls/gsnforwarder/scripts/generate-bindings.sh:33]: IForwarder.sol is included by Forwarder.sol, so this code wasn't generating anything that wasn't already generated.

File: examples/mock_recipient/mock_test.go [hunk 6]
```diff
@@ -26,72 +23,64 @@ var ganachePrivateKeys = []string{
...
-	return auth, ec, pk
-}
-
 func TestMock_Execute(t *testing.T) {
 	ctx, cancel := context.WithTimeout(context.Background(), time.Second*30)
 	defer cancel()
 
-	auth, conn, pk := setupAuth(t)
-	chainID, err := conn.ChainID(ctx)
+	pk, err := ethcrypto.HexToECDSA(ganachePrivateKeys[0])
```
Comment (dimalinux) [inline examples/mock_recipient/mock_test.go:30]: This test isn't using the gsnforwarder, and I ended up spending a lot of time to debug it, that could have been used for atomic swap issues. It was reusing the same txOpts on multiple transactions and it wasn't setting the nonce in the transfer.

File: impls/gsnforwarder/contracts/@openzeppelin/contracts/interfaces/IERC165.sol [hunk 7]
```diff
@@ -0,0 +1,6 @@
+// SPDX-License-Identifier: MIT
+// OpenZeppelin Contracts v4.4.1 (interfaces/IERC165.sol)
```
Comment (dimalinux) [inline impls/gsnforwarder/contracts/@openzeppelin/contracts/interfaces/IERC165.sol:2]: I was originally confused, thinking that they combined multiple OpenZeppelin versions, but it turns out that OpenZeppelin only updates the version when they change the contract. We still don't know what version they used, but it must be >= `v4.7.3`.

File: impls/gsnforwarder/contracts/forwarder/Forwarder.sol
Comment (dimalinux) [inline impls/gsnforwarder/contracts/forwarder/Forwarder.sol:1]: The 2 main contracts were moved into a `forwarder` folder to match the paths shown for the files on etherscan. The path to the files relative to `solc` affect the metadata. Here's how you can download the sources:
```
curl 'https://api.etherscan.io/api?module=contract&action=getsourcecode&address=0xB2b5841DBeF766d4b521221732F9B618fCf34A87' > sourcecode.json
```
It's not in a very parse-friendly format. Someone should write a polished tool to parse out the files and write them to a directory. The code I wrote was far too dirty to keep.

File: impls/gsnforwarder/scripts/generate-bindings.sh [hunk 8]
```diff
@@ -1,36 +1,41 @@
...
+	--workdir /contracts
+	-u "$(id -u):$(id -g)"
+	"ethereum/solc:${SOLC_VERSION}"
+)
+
+SOLC_EXEC+=(--allow-paths . --optimize)
+
+set -x
+"${SOLC_EXEC[@]}" --abi forwarder/Forwarder.sol -o abi/ --overwrite
+"${SOLC_EXEC[@]}" --bin forwarder/Forwarder.sol -o bin/ --overwrite
```
Comment (dimalinux) [inline impls/gsnforwarder/scripts/generate-bindings.sh:30]: For using a very specific, older version of solc, using docker seems like the best way to go.

**Review by noot — APPROVED** [review]
looks good, nice work! is it easy to change the solidity versions with docker? might be nice to document

**Comment by dimalinux** [comment]
> looks good, nice work! is it easy to change the solidity versions with docker?

Yes, in that script there is just a single variable for the solidity version. It is the "tag" portion (the part after the colon) of the docker image. The flags that solc accepts have changed with time, so if you are using special flags or newer flags, there could be more involved than just changing the version.

**Pull request merged**